### PR TITLE
Cache number of topic-partitions across messages in a batch

### PR DESCRIPTION
**Pull request by pralav-kloudfuse**
Cache number of topic-partitions across messages in a batch

### CODE REVIEW

File: writer.go [hunk 1]
```diff
@@ -637,15 +637,21 @@ func (w *Writer) WriteMessages(ctx context.Context, msgs ...Message) error {
...
 		topic, err := w.chooseTopic(msg)
 		if err != nil {
 			return err
 		}
 
-		numPartitions, err := w.partitions(ctx, topic)
-		if err != nil {
-			return err
+		numPartitions, exists := topicPartitionCount[topic]
+		if !exists {
```
Comment (ashishkf) [inline writer.go:649]: you can combine above two lines - if <line 648>; !exists { ... }

**Review by ashishkf — APPROVED** [review]

File: writer.go (hunk 1 above)
Comment (pralav-kloudfuse) [reply writer.go:649]: when i do that, Go doesn't like the numPartitions i have outside the if statement (at 657)

**Review by pankaj-kf — COMMENTED** [review]
Is this a cross port of some upstream PR? Did this show significantly in the profile?

**Comment by pralav-kloudfuse** [comment]
@pankaj-kf we saw some optimization opportunities looking at the profile. Will do some testing to see if there is significant impact.

Screenshot 2023-11-23 at 10 02 08 AM

**Comment by pankaj-kf** [comment]
Not sure if this is worth it; lets not make changes to external library unless the gains are significant.

**Pull request closed without merging**